build.mk

```diff
@@ -18,7 +18,6 @@ GO := $(GO_CMD) $(GOFLAGS)
 all: build test ci_report checks # keep checks last as license test is brittle ...
 
 .PHONY: test
-test: build
```

Comment:
Not sure this is going in the right direction. The dependency allows to rebuild after changes.
Actually, we should make sure that the dependencies work, namely: they force a rebuild when there are changes; they do not rebuild anything if there are no changes.

Reply:
I agree for all cpp related subdirectories, this is the correct thing todo. However, go tests work differently as they do not require to run `go build` before `go test`. In particular, I removed this dependency here on the top level as all makefiles are including `build.mk` and thereby inherent this dependency. I suggest that we add this dependency where needed, eg. in `common/` and `ecc_enclave`

Reply:
I just checked `common/crypto` and `common/logging` already have a `build` dependency on `test`. `ecc_enclave` has currently no unit tests defined via cmake.

Reply:
another reason why I don't like the dependency defined in `build.mk` is that a `make` triggered from the root dir will first go through all subdirs and call `make build`, once this is done, it goes through all subdirs again calling `make test`, which will call `build` dependency in each subdir.

We currently also have certain subfolders where `make build` is not smart enough (or our way how to use make) to detect that nothing has changed and just runs the build again. Proto generate is an example.

Reply:
>another reason why I don't like the dependency defined in build.mk is that a make triggered from the root dir will first go through all subdirs and call make build, once this is done, it goes through all subdirs again calling make test, 
which will call build dependency in each subdir.

Per se, this is not an issue. Ideally, the second `make build`  would simply return with a "nothing to do".

> We currently also have certain subfolders where make build is not smart enough (or our way how to use make) to detect that nothing has changed and just runs the build again. Proto generate is an example.

Yes, I agree. Even in the CI, the `make test` causes (for example) `libshim` to be rebuilt. So, in general, there is some duplication of build -- even if you just run `make build` twice.

I am fine removing it for now but, again, it's `make build` that should be made smarter.
Regarding the go part, avoiding the go build/test dependency (assuming we restore the build/test dependency) adds unnecessary complexity since the go build part takes less than a handful of seconds, and in the end you always have to build for end-to-end tests.